library/std/src/sys/sgx/abi/usercalls/alloc.rs

```diff
@@ -404,6 +417,106 @@ pub(crate) unsafe fn copy_to_userspace(src: *const u8, dst: *mut u8, len: usize)
     }
 }
 
+/// Copies `len` bytes of data from userspace pointer `src` to enclave pointer `dst`
+///
+/// This function mitigates AEPIC leak vulnerabilities by ensuring all reads from untrusted memory are 8-byte aligned
+///
+/// # Panics
+/// This function panics if:
+///
+/// * The `src` pointer is null
+/// * The `dst` pointer is null
+/// * The `src` memory range is not in user memory
+/// * The `dst` memory range is not in enclave memory
+///
+/// # References
+///  - https://www.intel.com/content/www/us/en/security-center/advisory/intel-sa-00657.html
+///  - https://www.intel.com/content/www/us/en/developer/articles/technical/software-security-guidance/advisory-guidance/stale-data-read-from-xapic.html
+pub(crate) unsafe fn copy_from_userspace(src: *const u8, dst: *mut u8, len: usize) {
+    // Copies memory region `[src; src + len[` to the enclave at `dst`. The source memory region
+    // is:
+    //  - strictly less than 8 bytes in size and may be
+    //  - located at a misaligned memory location
+    fn copy_misaligned_chunk_to_enclave(src: *const u8, dst: *mut u8, len: u8) {
+        let mut tmp_buff = [0u8; 16];
+
+        // Compute an aligned memory region to read from
+        // +--------+ <-- aligned_src + aligned_len (8B-aligned)
+        // |  pad1  |
+        // +--------+ <-- src + len (misaligned)
+        // |        |
+        // |        |
+        // |        |
+        // +--------+ <-- src (misaligned)
+        // |  pad0  |
+        // +--------+ <-- aligned_src (8B-aligned)
+        let pad0_size = src as usize % 8;
+        let aligned_src = (src as usize - pad0_size) as _;
+
+        let pad1_size = 8 - (src as usize + len as usize) % 8;
+        let aligned_len = pad0_size + len as usize + pad1_size;
+
+        debug_assert!(len < 8);
+        debug_assert_eq!(aligned_src as usize % 8, 0);
+        debug_assert_eq!(aligned_len % 8, 0);
+        debug_assert!(aligned_len < 16);
```

Comment:
See discussion above.

Reply:
Then this should check `aligned_len <= 16`, right?

Reply:
Oops. Good catch! I've updated the commit.

I would've imagined that the `test_copy_from_userspace_function` would've tripped this. Aren't these `debug_assert` statements evaluated during those tests?

Reply:
AFAIK, CI isn't running *any* SGX tests, since it's a tier-2 target. If you tried those tests yourself, the debug-asserts may have been compiled out since `std` is usually compiled in release mode. Is it worth turning these into full-time asserts? LLVM might optimize them out anyway if it can statically prove them.

Reply:
I did run these tests locally. It must indeed be because stdlib was build in release mode. That's good to know! I don't think it's worth turning these `debug_asserts` into `asserts` and pay the performance penalty.